### PR TITLE
Zoho Provider implementation

### PR DESCRIPTION
**Pull request by smiziara**
## Description

Implementation of the Zoho OAuth provider.

## Motivation and Context

We use Zoho products so we need this provider

## How Has This Been Tested?

Tested on Centos 7 server and currently being used on production.

## Checklist:

- [X] My change requires a change to the documentation or CHANGELOG.
- [X] I have updated the documentation/CHANGELOG accordingly.
- [X] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by smiziara** [comment]
@JoelSpeed i tryed the OIDC but could not make it work as intended. Zoho doesn't support open id and i need to get the email address from the token.
Should i retry using Oidc in a different way?

**Comment by JoelSpeed** [comment]
@smiziara I think you're right. If they do not support OpenID in the scope then you will indeed need a separate provider for this. I will review what you've done 

@syscll @steakunderscore WDYT of this? Also, we really really need to set up a better way to handle providers to make this easier

File: providers/zoho.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+// NewZohoProvider initiates a new ZohoProvider
+func NewZohoProvider(p *ProviderData) *ZohoProvider {
+	p.ProviderName = "Zoho"
+	if p.LoginURL.String() == "" {
+		p.LoginURL = &url.URL{Scheme: "https",
+			Host:     "accounts.zoho.com",
+			Path:     "/oauth/v2/auth",
+			RawQuery: "response_type=code&access_type=offline",
```
Comment (JoelSpeed) [inline providers/zoho.go]: I've not seen `RawQuery` set in one of the providers before, I noticed that this duplicates the call to set `response_type` [here](https://github.com/pusher/oauth2_proxy/blob/9670f54dd00fd75ffd0fb765f8fd60aa64c1fabd/providers/provider_default.go#L92). But the `access_type` is new. Can you verify for me that all the parameters are set correctly in by generating a login URL?  Is there a test for the login URL?

File: providers/zoho.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	}
+	return &ZohoProvider{ProviderData: p}
+}
+
+func getZohoHeader(accessToken string) http.Header {
+	header := make(http.Header)
+	header.Set("Content-Type", "application/json")
+	header.Set("Authorization", fmt.Sprintf("Bearer %s", accessToken))
+	return header
+}
```
Comment (JoelSpeed) [inline providers/zoho.go]: This is identical to [getOIDCToken](https://github.com/pusher/oauth2_proxy/blob/master/providers/oidc.go#L175-L180), perhaps that could be re-used?

Comment (smiziara) [reply providers/zoho.go]: @JoelSpeed the response type is indeed duplicated, i will fix. The access type needs to be sent for zoho. Online means only access token is sent and lasts for 1 hour. Offline, they send a refresh token.

About reusing methods directly from different strategies, IMHO its not a good ideia, but i can change it if you want. 

Maybe we can create an abstract oauth provider that can be flexible to setup from config files.

Comment (JoelSpeed) [reply providers/zoho.go]: > Offline, they send a refresh token.

This is common to many providers, I wonder why this isn't being set anywhere else 🤔 

> About reusing methods directly from different strategies, IMHO its not a good ideia, but i can change it if you want.

Happy to leave it, thanks for pushing back :) 

> Maybe we can create an abstract oauth provider that can be flexible to setup from config files.

Yes x1000!!! I have been thinking about this for months, trying to work out a good way to do this and transition the existing providers is a problem for me though

**Pull request closed without merging**